dev-tools/mage/config.go

```diff
@@ -18,16 +18,150 @@
+)
+
+// ConfigFileType is a bitset that indicates what types of config files to
+// generate.
+type ConfigFileType uint8
+
+// Config file types.
+const (
+	ShortConfigType ConfigFileType = 1 << iota
+	ReferenceConfigType
+	DockerConfigType
+
+	AllConfigTypes ConfigFileType = 0xFF
+)
+
+// IsShort return true if ShortConfigType is set.
+func (t ConfigFileType) IsShort() bool { return t&ShortConfigType > 0 }
+
+// IsReference return true if ReferenceConfigType is set.
+func (t ConfigFileType) IsReference() bool { return t&ReferenceConfigType > 0 }
+
+// IsDocker return true if DockerConfigType is set.
+func (t ConfigFileType) IsDocker() bool { return t&DockerConfigType > 0 }
+
+// ConfigFileParams defines the files that make up each config file.
+type ConfigFileParams struct {
+	ShortParts     []string // List of files or globs.
+	ReferenceParts []string // List of files or globs.
+	DockerParts    []string // List of files or globs.
+	ExtraVars      map[string]interface{}
+}
+
+// Config generates config files. Set DEV_OS and DEV_ARCH to change the target
+// host for the generated configs. Defaults to linux/amd64.
+func Config(types ConfigFileType, args ConfigFileParams, targetDir string) error {
+	if err := makeConfigTemplates(types, args); err != nil {
+		return errors.Wrap(err, "failed making config templates")
+	}
+
+	params := map[string]interface{}{
+		"GOOS":      EnvOr("DEV_OS", "linux"),
+		"GOARCH":    EnvOr("DEV_ARCH", "amd64"),
+		"Reference": false,
+		"Docker":    false,
+	}
+	for k, v := range args.ExtraVars {
+		params[k] = v
+	}
+
+	// Short
+	if types.IsShort() {
+		file := filepath.Join(targetDir, BeatName+".yml")
+		fmt.Printf(">> Building %v for %v/%v\n", file, params["GOOS"], params["GOARCH"])
+		if err := ExpandFile(shortTemplate, file, params); err != nil {
+			return errors.Wrapf(err, "failed building %v", file)
+		}
+	}
+
+	// Reference
+	if types.IsReference() {
+		file := filepath.Join(targetDir, BeatName+".reference.yml")
+		params["Reference"] = true
+		fmt.Printf(">> Building %v for %v/%v\n", file, params["GOOS"], params["GOARCH"])
+		if err := ExpandFile(referenceTemplate, file, params); err != nil {
+			return errors.Wrapf(err, "failed building %v", file)
+		}
+	}
+
+	// Docker
+	if types.IsDocker() {
+		file := filepath.Join(targetDir, BeatName+".docker.yml")
+		params["Reference"] = false
+		params["Docker"] = true
+		fmt.Printf(">> Building %v for %v/%v\n", file, params["GOOS"], params["GOARCH"])
+		if err := ExpandFile(dockerTemplate, file, params); err != nil {
+			return errors.Wrapf(err, "failed building %v", file)
+		}
+	}
+
+	return nil
+}
+
+func makeConfigTemplates(types ConfigFileType, args ConfigFileParams) error {
+	var err error
+
+	if types.IsShort() {
+		if err = makeConfigTemplate(shortTemplate, 0600, args.ShortParts...); err != nil {
+			return err
+		}
+	}
+
+	if types.IsReference() {
+		if err = makeConfigTemplate(referenceTemplate, 0644, args.ReferenceParts...); err != nil {
```

Comment:
This matches what mode we use for packaging. The short config has really strict privs (0600) because the beat won't start if others can read/write it. The reference config's mode is more liberal since it's just for reference and contains no sensitive info.